baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/StreamFlatMapOptional.java

```diff
@@ -30,50 +32,146 @@
+            MatchType.AT_LEAST_ONE,
+            MethodMatchers.instanceMethod()
+                    .onDescendantOf(Optional.class.getCanonicalName())
+                    .namedAnyOf("get", "orElseThrow")
+                    .withNoParameters());
 
     @Override
     public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
         if (STREAM_FLATMAP_OPTIONAL_STREAM.matches(tree, state)) {
-            ExpressionTree receiver = ASTHelpers.getReceiver(tree.getMethodSelect());
-            if (receiver != null) {
-                SuggestedFix.Builder fix = SuggestedFix.builder();
-                String optionalType = SuggestedFixes.qualifyType(state, fix, Optional.class.getCanonicalName());
-                String replacement = ".filter(" + optionalType + "::isPresent).map(" + optionalType + "::get)";
-                return buildDescription(tree)
-                        .addFix(fix.replace(state.getEndPosition(receiver), state.getEndPosition(tree), replacement)
-                                .build())
-                        .build();
+            ExpressionTree methodSelect = tree.getMethodSelect();
+            ExpressionTree receiver = ASTHelpers.getReceiver(methodSelect);
+            return fix(tree, state, receiver, receiver);
+        }
+
+        if (STREAM_MAP_GET.matches(tree, state)) {
+            ExpressionTree mapTree = ASTHelpers.getReceiver(tree.getMethodSelect());
+            if (mapTree != null && STREAM_FILTER_IS_PRESENT.matches(mapTree, state)) {
+                ExpressionTree filterTree = ASTHelpers.getReceiver(mapTree);
+                return fix(tree, state, tree.getMethodSelect(), filterTree);
             }
         }
+
         return Description.NO_MATCH;
     }
+
+    private Description fix(
+            MethodInvocationTree tree,
+            VisitorState state,
+            @Nullable ExpressionTree receiver,
+            @Nullable ExpressionTree expressionTree) {
+        if (receiver == null || expressionTree == null) {
+            return Description.NO_MATCH;
+        }
+
+        Type elementType = ASTHelpers.getType(tree);
+        if (elementType == null) {
+            return Description.NO_MATCH;
+        }
+
+        Type receiverType = ASTHelpers.getType(receiver);
+        if (receiverType == null) {
+            return Description.NO_MATCH;
+        }
+
+        return Stream.<Supplier<SuggestedFix>>of(
+                        () -> getSuggestedFix(SuggestedFix.builder(), tree, state, expressionTree, ""),
+                        () -> getQualifiedSuggestedFix(tree, state, expressionTree, receiverType.getTypeArguments()),
+                        () -> getQualifiedSuggestedFix(tree, state, expressionTree, List.of(elementType)))
+                .<SuggestedFix>mapMulti((f, consumer) -> {
+                    SuggestedFix fix = f.get();
+                    if (SuggestedFixes.compilesWithFix(fix, state)) {
+                        consumer.accept(fix);
```

Comment:
I think we should guarantee that we propose _something_ after checking `compilesWithFix`, this could make future compilations expensive if compilation doesn't succeed on any of the options